File: dynamic-programming/climbing-stairs/climbing-stairs.ts
function climbStairs(n: number): number {
  if (n === 1) {
    return 1;
  }

  const dp: number[] = new Array(n + 1).fill(0);
  dp[1] = 1;
  dp[2] = 2;

  for (let i = 3; i <= n; i++) {
    dp[i] = dp[i - 1] + dp[i - 2];
  }

  return dp[n];
}

function climbStairs2(n: number): number {
  if (n === 1) {
    return 1;
  }

  let first = 1;
  let second = 2;

  for (let i = 3; i <= n; i++) {
    let third = first + second;
    first = second;
    second = third;
  }

  return second;
}

File: dynamic-programming/climbing-stairs/climbing-stairs.go
package climbingstairs

func climbStairs(n int) int {
	if n == 1 {
		return 1
	}

	dp := make([]int, n+1)

	dp[1] = 1
	dp[2] = 2

	for i := 3; i <= n; i++ {
		dp[i] = dp[i-1] + dp[i-2]
	}

	return dp[n]
}

func climbStairs2(n int) int {
	if n == 1 {
		return 1
	}

	first := 1
	second := 2

	for i := 3; i <= n; i++ {
		third := first + second
		first = second
		second = third
	}

	return second
}
